### PR TITLE
Minor extension to generic ip discovery

### PR DESCRIPTION
**Pull request by holgerfriedrich**
Add **new replacement $uuid** in request frame.
This is needed for discovery of ipcamera devices.

Add **new parameter listenPort** (required for govee).

Add **new parameter requestPlain** as alternative to request (enhance readability of request for text/xml/json based req., govee and ipcamera)

Improve resource handling:
* autocloseable
* ~~restrict to 5 addresses~~ (moved to next PR)
* ~~respect system.config.network.useOnlyOneAddress setting~~ (moved to next PR)

Refs: #3936

### CODE REVIEW

**Comment by holgerfriedrich** [comment]
@Skinah FYI

File: bundles/org.openhab.core.config.discovery.addon.ip/src/main/java/org/openhab/core/config/discovery/addon/ip/IpAddonFinder.java [hunk 1]
```diff
@@ -164,6 +177,22 @@ private void scan() {
...
+                int listenPort = 0; // default, pick a non-privileged port
+                if (parameters.get(PARAMETER_LISTEN_PORT) != null) {
+                    try {
+                        listenPort = Integer.parseInt(Objects.toString(parameters.get(PARAMETER_LISTEN_PORT)));
+                    } catch (NumberFormatException e) {
+                        logger.warn("{}: discovery-parameter '{}' cannot be parsed", candidate.getUID(),
+                                PARAMETER_LISTEN_PORT);
+                        continue;
+                    }
+                    // not not allow privileged ports
```
Comment (Skinah) [inline bundles/org.openhab.core.config.discovery.addon.ip/src/main/java/org/openhab/core/config/discovery/addon/ip/IpAddonFinder.java:189]:
```suggestion
                    // Do not allow privileged ports
```

**Comment by Skinah** [comment]
Thank you for the changes.

**Comment by andrewfg** [comment]
@holgerfriedrich the ZWay binding is probably a candidate to be suggested via your IP addon finder. => WDYT?

https://github.com/openhab/openhab-addons/tree/main/bundles/org.openhab.binding.zway#discovery

**Comment by holgerfriedrich** [comment]
@andrewfg I can support / test if we go for zwave instead...

**Comment by holgerfriedrich** [comment]
> @holgerfriedrich the ZWay binding is probably a candidate to be suggested via your IP addon finder. => WDYT?
> 
> https://github.com/openhab/openhab-addons/tree/main/bundles/org.openhab.binding.zway#discovery

Sorry, @andrewfg, I got you wrong in my first reply. Let's discuss this #3936. I will copy your comment to the other issue.

**Comment by holgerfriedrich** [comment]
@mherwege I think this PR is ready for review. It contains extensions needed for govee and ipcamera bindings. @stefan-hoehn just confirmed it working for govee and approved my PR openhab/openhab-addons#16109.

**Comment by andrewfg** [comment]
> support / test if we go for zwave instead...

Just for the avoidance of doubt:
- The Z-Wave binding is based on a Z-Wave USB stick attached to the PC which will be suggested by my USB finder.
- The ZWay binding is based on a Z-Wave bridge server somewhete on the LAN which would be suggested by your IP finder.

**Comment by holgerfriedrich** [comment]
@mherwege Could you please take a look at this?
The PR contains minor extensions needed for govee and ipcamera.
Additionally, improves the resource behavior (autocloseable DatagramChannel, limit to 5 interfaces, respect system.config.network.useOnlyOneAddress).

**Comment by holgerfriedrich** [comment]
@openhab/core-maintainers could you pls have a look? Auto-detect for 2 add-ons is depending on additional functions introduced in this PR:
- openhab/openhab-addons#16109 (govee)
- openhab/openhab-addons#16090 (ipcamera)

Thanks a lot!

**Comment by mherwege** [comment]
In principle, this LGTM.

But I am wondering if we should not introduce use of https://github.com/openhab/openhab-core/pull/3981 at the same time. As we add requests being sent out to all interfaces, we should probably restrict the interfaces before getting too much into trouble.

**Comment by holgerfriedrich** [comment]
@mherwege thanks for looking into this.

> But I am wondering if we should not introduce use of #3981 at the same time. As we add requests being sent out to all interfaces, we should probably restrict the interfaces before getting too much into trouble.

I had implemented a fallback to the main interface in case too many interfaces were found, but removed it for now to keep the PR cleaner.
I am in doubt that we would see an issue here because the requests are strictly sequential: we sent only one frame every few seconds (timeout is currently set to 5sec). 

What we will need to think about - but for me it is not yet scope of this PR - is either shortening the timeouts or doing requests more or less in parallel. Otherwise we will not finish the scan before the wizard suggests plugins during when OH is installed.

**Comment by mherwege** [comment]
> I had implemented a fallback to the main interface in case too many interfaces were found, but removed it for now to keep the PR cleaner.

Agree to keep it out of here, but that’s where it should probably be in sync with the work on https://github.com/openhab/openhab-core/pull/3981, so for a follow-up PR.



> What we will need to think about - but for me it is not yet scope of this PR - is either shortening the timeouts or doing requests more or less in parallel. Otherwise we will not finish the scan before the wizard suggests plugins during when OH is installed.

Maybe define CompletableFuturs with timeout for each request and start them in parallel. Agree we will need more throughput on this. We can delay initial configuration in the UI a bit, showing a preparing screen, within reason.

**Comment by holgerfriedrich** [comment]
@mherwege @andrewfg should we go forward merging this PR as it is?
Obviously, there is more work to do as described in #3936. This is just a first step.

**Comment by andrewfg** [comment]
I would suggest to use CompletableFuture as @mherwege suggests. Then the scans can all be executed in parallel on separate threads.

**Comment by mherwege** [comment]
> I would suggest to use CompletableFuture as @mherwege suggests. Then the scans can all be executed in parallel on separate threads.

I tend to agree here. It now becomes easy to add discovery info to add-ons. If a number of them are added in the addon repo, suddenly we have code running a sequence of scans potentially each timing out. We should avoid having to come back to this quickly when discovery info gets added to the addon.xml files.

**Comment by andrewfg** [comment]
> suddenly we have code running a sequence of scans potentially each timing out

In particular in OH core, there should be no long running processes on main threads.

**Comment by andrewfg** [comment]
I am thinking that the approach below would perhaps be appropriate..

https://github.com/openhab/openhab-addons/pull/16315

**Review by andrewfg — COMMENTED** [review]
@holgerfriedrich apropos threading .. to be a bit more helpful I would suggest refactoring your code lines 265..395 into a separate method, and at line #264 call that method via something like List<Future> myFutures .. add(scheduler.submit(() -> thatMethod()))`.

And on shutdown you should also do `myFutures.forEach(f -> cancel())`;

File: bundles/org.openhab.core.config.discovery.addon.ip/src/main/java/org/openhab/core/config/discovery/addon/ip/IpAddonFinder.java [hunk 2]
```diff
@@ -322,6 +396,30 @@ private void scan() {
         logger.trace("IpAddonFinder::scan completed");
     }
 
+    // build from plaintext string
+    private byte[] buildRequestArrayPlain(DatagramChannel channel, String request)
+            throws java.io.IOException, ParseException {
+        InetSocketAddress sock = (InetSocketAddress) channel.getLocalAddress();
+
+        // replace first
+        StringBuffer req = new StringBuffer(request);
```
Comment (wborn) [inline bundles/org.openhab.core.config.discovery.addon.ip/src/main/java/org/openhab/core/config/discovery/addon/ip/IpAddonFinder.java:405]: The `StringBuilder` is preferred when there is no need for synchronization as it has better performance:

```suggestion
        StringBuilder req = new StringBuilder(request);
```

**Comment by holgerfriedrich** [comment]
I agree that better handling of multiple addresses and interfaces is desired for IP discovery. Which solution to implement is not yet clear, maybe there will be a result out of #4047.

Parallelization is required as well when more add-ons start using this. Thanks, @mherwege and @andrewfg for pointing at ```CompletableFuture``` and implementations. There are some technical questions not yet clarified, e.g. how to make sure that discovery of different add-ons at the same time does not cause problems. Just imagine two add-ons using the same server port, etc...

I still think that both is out of scope of this PR and tracked #3936.

The PR adds new config options and has already been validated with the corresponding add-on.

**Comment by andrewfg** [comment]
@mherwege [HERE](
https://github.com/andrewfg/openhab-core/blob/ip-finder-async/bundles/org.openhab.core.config.discovery.addon.ip/src/main/java/org/openhab/core/config/discovery/addon/ip/IpAddonFinder.java) is some code to do the scans asynchronously (not tested).

**Review by wborn — APPROVED** [review]
Thanks for the extension!

**Comment by andrewfg** [comment]
@holgerfriedrich shall I convert my code to a new PR?

**Comment by holgerfriedrich** [comment]
> @holgerfriedrich shall I convert my code to a new PR?

Thanks, @andrewfg. I have created #4094 based on your example above. Hope you are fine with that.

**Pull request merged**